packages/fxa-settings/src/lib/relier-factory.ts

```diff
@@ -0,0 +1,176 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Context } from "./context/interfaces/context";
+import { RelierData, Relier } from "../models/reliers/rellier"
+import { BrowserRelier } from "../models/reliers/browser-relier";
+import { OAuthRelier } from "../models/reliers/oauth-relier";
+import { RelierPairingSupplicant } from "../models/reliers/pairing-supplicant-relier";
+import { PairingAuthorityRelier } from "../models/reliers/pairing-authority-relier";
+import { UrlContext } from "./context/implementations/url-context";
+import { StorageContext } from "./context/implementations/storage-context";
+
+const DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP = /^\/pair\/supp/;
+const DEVICE_PAIRING_AUTHORITY_REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob:pair-auth-webchannel';
+const SYNC_SERVICE = 'sync';
+const FX_DESKTOP_V3_CONTEXT = 'fx_desktop_v3'
+
+/**
+ * A react context compliant interface
+ */
+export interface IRelierContext {
+  relier: RelierData,
+  setRelier: (relier:RelierData) => void
+};
+
+/**
+ * Creation flags interface, controls the type of relier that is ultimately produced.
+ */
+export interface RelierFlags
+{
+  isDevicePairingAsAuthority():boolean;
+  isDevicePairingAsSupplicant():boolean;
+  isOauth():boolean;
+  isSyncService():boolean;
+  isV3DesktopContext():boolean;
+}
+
+/**
+ * Produces Reliers
+ */
+export class RelierFactory
+{
+  /**
+   * Produces a relier given the current context and flags extrapolated from that context.
+   * @param context A key value data store that holds the relier state
+   * @param flags A set of flags that help determine what kind of relier to produce. These flags are also based off the context.
+   * @returns A relier implementation.
+   */
+  static getRelier(
+    context:Context,
+    flags:RelierFlags
+  ) {
+
+    // Keep trying until something sticks
+    if (flags.isDevicePairingAsAuthority()) {
+      return new PairingAuthorityRelier(context);
+    }
+    if (flags.isDevicePairingAsSupplicant()) {
+      return new RelierPairingSupplicant(context);
+    }
+    if (flags.isOauth()) {
+      return new OAuthRelier(context);
+    }
+    if (flags.isSyncService() && flags.isV3DesktopContext()) {
```

Comment:
Does this need to be `||` instead of `&&`? I see this comment:

```
...
      } else if (
        this._isServiceSync() ||
        // context v3 is able to sign in to Firefox without enabling Sync
        this._searchParam('context') === Constants.FX_DESKTOP_V3_CONTEXT
      ) {
...
```

Reply:
Nice catch, thank you.